compiler/rustc_passes/src/dead.rs

```diff
@@ -976,6 +976,16 @@ impl<'tcx> DeadVisitor<'tcx> {
         };
 
         let encl_def_id = parent_item.unwrap_or(first_item.def_id);
+        // If parent of encl_def_id is an enum, use the parent ID intead.
+        let encl_def_id = if let Some(enum_did) = self.tcx.opt_parent(encl_def_id.to_def_id())
+            && let Some(enum_local_id) = enum_did.as_local()
+            && let Node::Item(item) = self.tcx.hir_node_by_def_id(enum_local_id)
+            && let ItemKind::Enum(_, _) = item.kind
```

Comment:
This logic is making me nervous.

If I understand correctly, this is putting in a bit of an ad-hoc workaround, specifically into the dead-code lint pass, to hop up two parents instead of just one parent if the second parent is itself identified to be an enum.

i infer that is because it is trying to follow a line of reasoning like "the second parent will be an enum if *and only if* the first parent is an enum variant and the original item is a data field attached to that enum variant.

1. It is not immediately obvious to me that such an "if and only if" actually holds.

2. Even if it does hold, my second worry is that I imagine this same mistake probably presents itself in a bunch of other places (i.e., places where people assumed that the parent of a data field would be the ADT type that holds that field, but in fact one has to go up two parent-hops instead of one in cases like this), and I wonder if we would be better off using an API that let one express that intention directly, rather than putting logic like this everywhere it needs to go...

Reply:
> Even if it does hold, my second worry is that I imagine this same mistake probably presents itself in a bunch of other places

I've seen issues around the Enum/Variant/CTor trichotomy, where the `DefId` of the item is expected for some analysis but you get passed one of the other two, so I'm sure there are more lurking around.

Could you let us know what the contents of `parent_item` (what it points to, really) and of `first_item` are?

Reply:
I talked to @estebank independently.

While I do think better API's might exist, I don't think I want to block this PR on hypothetical API improvements.

What I *do* want from this PR is for it to change the code slightly to include a few more, potentially redundant, safe-guards where you have the code check, as it does its double hop, that each item kind it encounters is the one you expect (e.g. taken from Enum/Variant/Ctor as appropriate.)

@rustbot author

Reply:
Thank you for comment and sorry for the late reply.

I'd like to confirm my understanding. We need to check `encl_def_id` is a just `Variant` (or other kind?) because other item such as  `Ctor` also has `Enum` parent.

Reply:
yes that is what we're asking for.